### PR TITLE
Adjust quote based on Bitcoin balance

### LINKED ISSUE #939: Adjust quote based on Bitcoin balance
The maximum swap amount should be calculated based on the Balance of the internal bitcoin wallet. A swap failing after depositing Bitcoin leads to poor UX.

### PR DESCRIPTION
**Pull request by lescuer97**
Fixes #939 #963 
Please comment in the new method for getting a Monero value, I had to allow clippy::cast_precision_loss to convert to a f64, for our use case I don't really thing we will lose much precision. 

Please comment on the implementation of the check.

### CODE REVIEW

**Comment by binarybaron** [comment]
Thanks a lot for implementing this! I'll look over it and give some feedback where I can

File: swap/src/monero.rs [hunk 1]
```diff
@@ -98,6 +98,10 @@ impl Amount {
     pub fn as_piconero(&self) -> u64 {
         self.0
     }
+    #[allow(clippy::cast_precision_loss)]
```
Comment (thomaseizinger) [inline swap/src/monero.rs:101]: Instead of allowing the clippy lint, please use `Decimal` for the calculations.

Comment (thomaseizinger) [reply swap/src/monero.rs:101]: A test would be good for this function!

File: swap/src/asb/event_loop.rs [hunk 2]
```diff
@@ -318,16 +318,30 @@ where
...
             .context("Failed to get latest rate")?;
 
+        max_bitcoin_for_monero = bitcoin::Amount::from_btc(monero_balance * rate.ask()?.as_btc())?;
+        if max_buy > max_bitcoin_for_monero {
+            tracing::warn!(
+                    "Your Max bitcoin to sell: {} is smaller than your current balance of monero available to sell {}",
+                    max_buy, max_bitcoin_for_monero
+                );
+            max_bitcoin_buy = max_bitcoin_for_monero
+        }
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:339]: Please implement this without mutable variables, this is too hard to follow in what is being modified under which conditions.

File: swap/src/asb/event_loop.rs [hunk 3]
```diff
@@ -318,16 +318,30 @@ where
...
+
         Ok(BidQuote {
             price: rate.ask().context("Failed to compute asking price")?,
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:342]: Getting the `ask` price is duplicated here with line 332.

File: swap/src/asb/event_loop.rs [hunk 4]
```diff
@@ -318,16 +318,30 @@ where
...
+        let max_bitcoin_for_monero: bitcoin::Amount;
+        let mut max_bitcoin_buy = max_buy;
+        let monero_balance = monero_wallet.get_balance().await?.as_monero();
+
         let rate = self
             .latest_rate
             .latest_rate()
             .context("Failed to get latest rate")?;
 
+        max_bitcoin_for_monero = bitcoin::Amount::from_btc(monero_balance * rate.ask()?.as_btc())?;
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:332]: No floating point math please. Either operate on sats and piconeros (which are both u64 I think) or use `Decimal` if you can't use natural numbers.

File: swap/src/asb/event_loop.rs [hunk 5]
```diff
@@ -318,16 +318,30 @@ where
         &mut self,
         min_buy: bitcoin::Amount,
         max_buy: bitcoin::Amount,
+        monero_wallet: Arc<monero::Wallet>,
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:321]: No need to pass this as a parameter, you have it on `self`. (Also visible by the fact that you pass `self.monero_wallet` into this function!)

File: swap/src/asb/event_loop.rs (hunk 4 above)
Comment (thomaseizinger) [reply swap/src/asb/event_loop.rs:332]: Also, it is a bit of an anti-pattern to extract the raw value from a newtype (the `as_monero` function call) because the type system no longer helps you in correctly interpreting the result of the computation.

You should be able to express this functionality by implementing `Mul` for `Rate` and `Monero`.

File: swap/src/monero.rs (hunk 1 above)
Comment (thomaseizinger) [reply swap/src/monero.rs:101]: Also: Missing whitespace above :)

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thanks! Left some comments :)

File: swap/src/asb/event_loop.rs [hunk 6]
```diff
@@ -318,16 +318,30 @@ where
...
+            tracing::warn!(
+                    "Your Max bitcoin to sell: {} is smaller than your current balance of monero available to sell {}",
+                    max_buy, max_bitcoin_for_monero
+                );
+            max_bitcoin_buy = max_bitcoin_for_monero
+        }
+
         Ok(BidQuote {
             price: rate.ask().context("Failed to compute asking price")?,
             min_quantity: min_buy,
```
Comment (binarybaron) [inline swap/src/asb/event_loop.rs:343]: If `max_bitcoin_buy` is less than `min_buy`, the user cannot initiate a swap at all. We need to handle this case somehow. I suggest that we set both `min_quantity` and `max_quantity` to zero to communicate to the user that the ASB is not accepting swaps at the moment.

**Review by binarybaron — CHANGES_REQUESTED** [review]

File: swap/src/asb/event_loop.rs (hunk 6 above)
Comment (lescuer97) [reply swap/src/asb/event_loop.rs:343]: I think we could make another issue for the project so we can solve it in parts

File: swap/src/asb/event_loop.rs (hunk 4 above)
Comment (lescuer97) [reply swap/src/asb/event_loop.rs]: Thank you for the input! I ended up using decimals at the end. I think the UX for the BTC and XMR nomenclature is better than Piconero and Satoshi one

File: swap/src/asb/event_loop.rs [hunk 7]
```diff
@@ -319,13 +320,40 @@ where
...
+                tracing::warn!("Using Max purchase from config: {}", max_buy);
+                max_buy
+            }
+        };
+
+        if max_buy > max_bitcoin_for_monero {
+            tracing::warn!(
+                    "Your Max bitcoin to sell config: {} is bigger than your current balance of Monero available to sell {}",
+                    max_buy, max_bitcoin_for_monero
+                );
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:347]: Can you please extract this calculation into a dedicated function or even better, move it _onto_ the actual types itself?

There is a lot of low-level logic happening here and we don't have any tests for it.

File: swap/src/asb/event_loop.rs [hunk 8]
```diff
@@ -319,13 +320,40 @@ where
...
             .latest_rate()
-            .context("Failed to get latest rate")?;
+            .context("Failed to get latest rate")?
+            .ask()
+            .context("Failed to compute asking price")?;
+
+        let bitcoin_decimal: Decimal = Decimal::from(rate.as_sat()) / Decimal::from(100_000_000);
+
+        let max_bitcoin_for_monero = match (monero_decimal * bitcoin_decimal).to_f64() {
+            Some(val) => bitcoin::Amount::from_btc(val)?,
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:335]: Use `to_u64` instead and `from_sat` please. `from_btc` involves floating point arithmetic!

File: swap/src/monero.rs [hunk 9]
```diff
@@ -360,6 +370,22 @@ mod tests {
...
+        assert_eq!(Decimal::new(10, 0), monero_decimal);
+    }
+
+    #[test]
+    fn create_piconero_offset_decimal() {
+        let monero = "10";
+        let amount = Amount::parse_monero(monero).unwrap();
+        let monero_decimal = amount.as_monero_decimal();
+        assert_eq!(Decimal::new(10, 0), monero_decimal);
+    }
```
Comment (thomaseizinger) [inline swap/src/monero.rs:387]: The same test twice with a different name?

File: swap/src/monero.rs [hunk 10]
```diff
@@ -99,6 +99,16 @@ impl Amount {
...
 
+    pub fn as_monero_decimal(&self) -> Decimal {
+        let piconero_decimal = Decimal::from(self.0);
+        let pic_offcet_decimal = Self::piconero_offset_decimal();
+        piconero_decimal / pic_offcet_decimal
+    }
+
+    pub fn piconero_offset_decimal() -> Decimal {
+        Decimal::from(PICONERO_OFFSET)
+    }
```
Comment (thomaseizinger) [inline swap/src/monero.rs:110]: Does not need to be public AFAIK.

File: swap/src/asb/event_loop.rs (hunk 7 above)
Comment (thomaseizinger) [reply swap/src/asb/event_loop.rs:347]: If you look at what the function did before, then you can see that it only gathered some data and combined it into a `BidQuote` but the actual calculation of what an _ask_ price is, does not happen in this function.

It would like to it stay that way. In other words: The responsibility of `make_quote` should be to construct a `BidQuote` by gathering data from several dependencies (like wallets, rates etc). The actual calculation of the price is better extracted into a pure function that does the computation based on the inputs. That function we can then test for all its edge-cases.

File: swap/src/asb/event_loop.rs [hunk 11]
```diff
@@ -319,13 +320,40 @@ where
         min_buy: bitcoin::Amount,
         max_buy: bitcoin::Amount,
     ) -> Result<BidQuote> {
+        let monero_decimal = self.monero_wallet.get_balance().await?.as_monero_decimal();
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:323]: Nit: The life-time of this variable is unnecessarily long. It is only used from line 334 so there isn't really any reason why this is at the very top!

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thanks for incorporating the feedback!

A few more comments :)

File: swap/src/asb/event_loop.rs (hunk 6 above)
Comment (binarybaron) [reply swap/src/asb/event_loop.rs:357]: I don't think we need another PR for this. Afaik this is already done for when the actual swap is started. You just need to subtract the fee from the Monero balance when calculating the `max_quantity`

Comment (lescuer97) [reply swap/src/asb/event_loop.rs:357]: You are right, I have implemented a the calculation to subtract the fee

File: swap/src/asb/event_loop.rs (hunk 8 above)
Comment (lescuer97) [reply swap/src/asb/event_loop.rs:335]: I managed a way to use u64 but had to watch for an underflow for subtracting the Monero fee

File: swap/src/monero.rs (hunk 9 above)
Comment (lescuer97) [reply swap/src/monero.rs:387]: yeah sorry I actually made an erroneous commit

File: swap/src/asb/event_loop.rs (hunk 6 above)
Comment (binarybaron) [reply swap/src/asb/event_loop.rs:349]: Thanks a lot for implementing this fee subtraction thing, looks good to me!

The scenario I originally mentioned where the min_quantity is bigger than the max_quantity is still possible afaik (mistake on my end). I'd suggest than we just check if min_quantity > max_quantity and then return a Quote where both are zero.

If either of you has a better suggestion, I'm definitely open to it.

File: swap/src/monero.rs [hunk 12]
```diff
@@ -99,6 +99,31 @@ impl Amount {
...
+
+    pub fn max_bitcoin_from_balance(&self, satoshis: u64) -> Result<bitcoin::Amount> {
+        let piconero_minus_fee = self.as_piconero().saturating_sub(MONERO_FEE.as_piconero());
+
+        if piconero_minus_fee == 0 {
+            tracing::error!("The current Monero balance is smaller than the basic locking fee");
+            bail!("The current Monero balance is smaller than the basic locking fee");
+        }
```
Comment (binarybaron) [inline swap/src/monero.rs:118]: Wouldn't it be more logical to just return zero? When calling this function we are essential asking "How much Bitcoin can we convert this into at this exchange rate while accounting for the fee?". If the answer to that questions happens to be zero, that's not really an error imo.

Also, will the ASB return a quote at all when the balance is zero?

Comment (lescuer97) [reply swap/src/monero.rs:118]: the 0 is actually not the real number if I had to use a normal subtraction I would get a negative number but because its a `u64` integer I would get an integer underflow panic, so I returned this as an elegant error to the asb provider. 

For the swap app if you get request a quote from a rendezvous server you wont get one from the the particular ASB with the lower then minimum balance and if you request it directly the connection will drop.

File: swap/src/asb/event_loop.rs [hunk 13]
```diff
@@ -322,10 +322,30 @@ where
         let rate = self
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:322]: Nit: Perhaps this variable should now say `ask_price`?

File: swap/src/asb/event_loop.rs [hunk 14]
```diff
@@ -322,10 +322,30 @@ where
...
-            .context("Failed to get latest rate")?;
+            .context("Failed to get latest rate")?
+            .ask()
+            .context("Failed to compute asking price")?;
+
+        let max_bitcoin_for_monero = self
+            .monero_wallet
+            .get_balance()
+            .await?
+            .max_bitcoin_from_balance(rate.as_sat())?;
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:333]: Why not pass the `bitcoin::Amount` directly? Gives you more type-safety! :)

File: swap/src/monero.rs [hunk 15]
```diff
@@ -99,6 +99,31 @@ impl Amount {
         self.0
     }
 
+    pub fn as_monero_decimal(&self) -> Decimal {
```
Comment (thomaseizinger) [inline swap/src/monero.rs:102]: Do we still need this now?

File: swap/src/monero.rs [hunk 16]
```diff
@@ -99,6 +99,31 @@ impl Amount {
...
+        let piconero_decimal = Decimal::from(self.0);
+        let pic_offcet_decimal = Self::piconero_offset_decimal();
+        piconero_decimal / pic_offcet_decimal
+    }
+
+    fn piconero_offset_decimal() -> Decimal {
+        Decimal::from(PICONERO_OFFSET)
+    }
+
+    pub fn max_bitcoin_from_balance(&self, satoshis: u64) -> Result<bitcoin::Amount> {
```
Comment (thomaseizinger) [inline swap/src/monero.rs:112]: I think a more descriptive name would be:


```suggestion
    pub fn max_bitcoin_for_price(&self, ask_price: u64) -> Result<bitcoin::Amount> {
```

File: swap/src/monero.rs [hunk 17]
```diff
@@ -99,6 +99,31 @@ impl Amount {
...
+        let piconero_minus_fee = self.as_piconero().saturating_sub(MONERO_FEE.as_piconero());
+
+        if piconero_minus_fee == 0 {
+            tracing::error!("The current Monero balance is smaller than the basic locking fee");
+            bail!("The current Monero balance is smaller than the basic locking fee");
```
Comment (thomaseizinger) [inline swap/src/monero.rs:117]: Either log or return error please but not both! That will lead to duplicate information in the logs :)

File: swap/src/monero.rs [hunk 18]
```diff
@@ -99,6 +99,31 @@ impl Amount {
...
+        }
+
+        // There needs to be an offset for difference in zeroes beetween Piconeros and
+        // Satoshis
+        let piconero_calc = (piconero_minus_fee * satoshis) / 1_000_000_000_000;
```
Comment (thomaseizinger) [inline swap/src/monero.rs:122]: Where does this constant come from? Can this be made more descriptive?

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thank you! I think we are getting close to a good solution here. Some more comments :)

File: swap/src/monero.rs (hunk 15 above)
Comment (lescuer97) [reply swap/src/monero.rs:102]: not really, do I remove it or leave it be?

File: swap/src/monero.rs (hunk 17 above)
Comment (lescuer97) [reply swap/src/monero.rs:117]: Done!

File: swap/src/asb/event_loop.rs (hunk 14 above)
Comment (lescuer97) [reply swap/src/asb/event_loop.rs:333]: Done!

File: swap/src/monero.rs (hunk 15 above)
Comment (thomaseizinger) [reply swap/src/monero.rs:102]: Remove it then please!

In general, I'd prefer if we strive towards having only code around that is actively used and relevant :)

Comment (thomaseizinger) [reply swap/src/monero.rs:102]: For PRs in particular, it is easier to review if the diff is minimal.

Comment (lescuer97) [reply swap/src/monero.rs]: Done!

File: swap/src/monero.rs [hunk 19]
```diff
@@ -99,6 +99,20 @@ impl Amount {
         self.0
     }
 
+    pub fn max_bitcoin_for_price(&self, ask_price: bitcoin::Amount) -> Result<bitcoin::Amount> {
+        let piconero_minus_fee = self.as_piconero().saturating_sub(MONERO_FEE.as_piconero());
+
+        if piconero_minus_fee == 0 {
+            bail!("The current Monero balance is smaller than the basic locking fee");
```
Comment (binarybaron) [inline swap/src/monero.rs:106]: I don't know if it's really the best idea to drop the connection when the balance is too low. It could lead to unexpected behavior and confuse the user of swap-cli, since he won't get any real feedback, but only some cryptic timeout error.

I still think returning a quote where the `max_quantity` is zero is the best workaround.

What do you think? @thomaseizinger

**Review by binarybaron — COMMENTED** [review]
Great work, thanks for your effort!

File: swap/src/monero.rs (hunk 19 above)
Comment (thomaseizinger) [reply swap/src/monero.rs:106]: Thanks for bringing this to my attention!

Yes, I think for a start, this function knows too much about the context it is called in. We should just return 0 here if that is the result of the computation.

If we still want to fail the quote protocol then we should do that one layer up!

Can we have a test run with both implementations and compare the user output?

File: swap/src/monero.rs [hunk 20]
```diff
@@ -360,6 +374,27 @@ mod tests {
...
+    #[test]
+    fn geting_max_bitcoin_to_trade() {
+        let monero = "10";
+        let amount = Amount::parse_monero(monero).unwrap();
+        let bitcoin_price_sats = bitcoin::Amount::from_sat(382_900);
+        let monero_max_from_bitcoin = amount.max_bitcoin_for_price(bitcoin_price_sats).unwrap();
+        assert_eq!(
+            bitcoin::Amount::from_sat(3_828_988),
+            monero_max_from_bitcoin
+        );
```
Comment (thomaseizinger) [inline swap/src/monero.rs:386]: A code style nit:

Have you heard of "arrange - act - assert" when it comes to testing? It makes tests heaps more readable when we strive to:

- Have a single statement per block (arrange/act/assert) or at least minimise their size
- Use whitespace to separate those blocks

For example:


```suggestion
        let amount = Amount::parse_monero("10").unwrap();
        let bitcoin_price_sats = bitcoin::Amount::from_sat(382_900);

        let monero_max_from_bitcoin = amount.max_bitcoin_for_price(bitcoin_price_sats).unwrap();

        assert_eq!(
            bitcoin::Amount::from_sat(3_828_988),
            monero_max_from_bitcoin
        );
```

Comment (lescuer97) [reply swap/src/monero.rs:386]: No! actually I didn't I thought that the normal thing was to call the function inside the Assertion!. Thank you for bringing that up!

Comment (thomaseizinger) [reply swap/src/monero.rs:386]: I wouldn't call the function inside the assertion, that would merge the "act" and "assert" stage :)

Simply adding whitespace between them makes it visually easy to identify the relevant parts of the test.

Comment (lescuer97) [reply swap/src/monero.rs]: just changed the tests

File: swap/src/monero.rs (hunk 19 above)
Comment (lescuer97) [inline swap/src/monero.rs:106]: do you want me to return something like this?
```suggestion
             if piconero_minus_fee == 0 {
            return bitcoin::Amount::ZERO;
```

Comment (thomaseizinger) [reply swap/src/monero.rs:106]: Yep, that is better!

Comment (lescuer97) [reply swap/src/monero.rs]: I just uploaded the suggestion from @binarybaron . But I think this should be a temporary solution because we are communicating to the user there is min_ buy amount and that could bring confusion. 

If there is not enough Monero to trade maybe we could give the SWAP user a graceful disconnect with a message or something like that

File: swap/src/asb/event_loop.rs [hunk 21]
```diff
@@ -319,13 +319,33 @@ where
...
+            tracing::warn!(
+                    "Your Max bitcoin to sell config: {} is bigger than your current balance of Monero available to sell {}",
+                    max_buy, max_bitcoin_for_monero
+                );
+            return Ok(BidQuote {
+                price: ask_price,
+                min_quantity: min_buy,
+                max_quantity: max_bitcoin_for_monero,
+            });
+        }
```
Comment (binarybaron) [inline swap/src/asb/event_loop.rs:346]: If max_bitcoin_buy is less than min_buy, the user cannot initiate a swap at all. We need to handle this case somehow. I suggest that we set both min_quantity and max_quantity to zero to communicate to the user that the ASB is not accepting swaps at the moment. Otherwise there might be a weird case where for example `min_quantity = 1 BTC` and `max_quantity = 0.5 BTC` which doesn't really make sense.

Something like this:
```rust
        if min_buy > max_bitcoin_for_monero {
             tracing::warn!(
                     "Your Monero balance is too low to initiate a swap, as your minimum swap amount is {}. You could at most swap {}",
                     min_buy, max_bitcoin_for_monero
                 );
             return Ok(BidQuote {
                 price: ask_price,
                 min_quantity: bitcoin::Amount::ZERO,
                 max_quantity: bitcoin::Amount::ZERO,
             });
         }
```

**Review by binarybaron — CHANGES_REQUESTED** [review]

File: swap/src/asb/event_loop.rs (hunk 21 above)
Comment (lescuer97) [reply swap/src/asb/event_loop.rs:346]: I will wait for @thomaseizinger  response on the solution before upload, but it seems fine to me

Comment (thomaseizinger) [reply swap/src/asb/event_loop.rs:346]: Yep, seems reasonable to me too. How does the swap tool handle both amounts being 0? Did someone do a test run for this?

Comment (lescuer97) [reply swap/src/asb/event_loop.rs:346]: I didn't attempt a trade but this where the outputs that  it gave me. 
Screenshot from 2022-03-01 09-05-42-cropped

Comment (thomaseizinger) [reply swap/src/asb/event_loop.rs:346]: Can you attempt a trade? :)

Comment (lescuer97) [reply swap/src/asb/event_loop.rs:346]: Okeyy, this are the results: The first time I tried to make a swap I't actually gave me a qr code to deposit and I tried it, as soon as the bitcoin arrived it failed and I received this message: 
Screenshot from 2022-03-01 11-05-47-cropped

Then if I tried to do a swap again it wouldn't even let me swap it would just send the error again.
Screenshot from 2022-03-01 11-07-23

The Bitcoin from the first transaction redeemed correctly

Comment (thomaseizinger) [reply swap/src/asb/event_loop.rs:346]: Okay, so the behaviour here is that we try to do a swap with the bitcoin already stored in the wallet and don't have you deposit new coins.

I think together with this change, we should add a check to the client whether or not `maximum_amount` is `0` and if so, abort directly.

Comment (lescuer97) [reply swap/src/asb/event_loop.rs:358]: Done!

File: swap/src/bin/swap.rs [hunk 22]
```diff
@@ -556,6 +556,11 @@ where
 {
     tracing::debug!("Requesting quote");
     let bid_quote = bid_quote.await?;
+
+    if bid_quote.max_quantity == bitcoin::Amount::ZERO {
+        bail!("The Monero seller you contacted can't trade monero at the moment")
```
Comment (thomaseizinger) [inline swap/src/bin/swap.rs:561]: I know you wrote this from a user's perspective but in the context of this function, I'd like a different error message here. Ideally, we would create a custom error type like `struct ZeroQuote;` and return that and react to that outside of this function.

If you don't want to go that far, I'd suggest just bailing with something like:

```suggestion
        bail!("Received quote with `max_quantity` of 0");
```

And adding a `.context` to the call of `determine_btc_to_swap` here: https://github.com/comit-network/xmr-btc-swap/blob/8f3f4d5895409775506eb256f35e945da1e52de6/swap/src/bin/swap.rs#L110

This will add a "layer" to the error message that should give enough context to the user of what is going wrong. (Read about [anyhow](https://docs.rs/anyhow/latest/anyhow/index.html) if you are not familiar with it.)

File: swap/src/asb/event_loop.rs [hunk 23]
```diff
@@ -319,13 +319,45 @@ where
...
+            .monero_wallet
+            .get_balance()
+            .await?
+            .max_bitcoin_for_price(ask_price);
+
+        if min_buy > max_bitcoin_for_monero {
+            tracing::warn!(
+                        "Your Monero balance is too low to initiate a swap, as your minimum swap amount is {}. You could at most swap {}",
+                        min_buy, max_bitcoin_for_monero
+                    );
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:339]: Formatting is a bit off here. Have you run the formatter on this code? :)

File: swap/src/asb/event_loop.rs [hunk 24]
```diff
@@ -319,13 +319,45 @@ where
...
+            return Ok(BidQuote {
```
Comment (thomaseizinger) [inline swap/src/asb/event_loop.rs:340]: Nit: Would prefer a newline before `return` statements.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thanks!

Some more comments and a formatting issue but overall, this is looking pretty good! Excited to merge this :)

File: swap/src/asb/event_loop.rs (hunk 23 above)
Comment (lescuer97) [reply swap/src/asb/event_loop.rs:339]: No, I forgot to do the command

File: swap/src/bin/swap.rs (hunk 22 above)
Comment (lescuer97) [reply swap/src/bin/swap.rs:561]: The structure of the Custom error how would you do it? with the same structure as BidQuote?

Comment (thomaseizinger) [reply swap/src/bin/swap.rs:561]: In a nutshell, it works like this:

First, we need to make the high-level decision of the function signature:

- Do we want to use `anyhow::Result` or `std::Result`? `anyhow` provides a boxed error type which can hold anything that implements `std::error::Error`. That is convenient but reduces type-safety a bit because we no longer know exactly, which errors are returned from the function.

At the moment, we use `anyhow::Result` and I'd suggest to keep it just to keep the code-churn of this PR low. If you research about `anyhow` and Rust error handling in general, you should find a few good resources and opinions.

With `anyhow::Error` (the error type of `anyhow::Result`), it is not quite as obvious how to do type-safe error handling. What we can do is:

1. Define a new error struct: `struct ZeroQuoteReceived;` that `derive`s `thiserror::Error` or implements `std::error::Error` manually.
2. Use `bail!(ZeroQuoteReceived)` at the above place.
3. On the call-site of `determine_btc_to_swap`, we can `match` on the returned `Result` (instead of `?` it) and then use the [`anyhow::Error::downcast` API](https://docs.rs/anyhow/latest/anyhow/struct.Error.html#method.downcast) to check whether the returned `anyhow::Error` "contains" a the `ZeroQuoteReceived` error type. If the `downcast` returns `Ok`, you can then handle that specific case by printing a nice message to the user. In the `Err` case, we can just `return` the error which is equivalent to what the current `?` does.

Comment (lescuer97) [reply swap/src/bin/swap.rs]: nice!, didn't knew about downcast

File: swap/src/bin/swap.rs [hunk 25]
```diff
@@ -99,15 +99,24 @@ async fn main() -> Result<()> {
...
                 max_givable,
                 || bitcoin_wallet.sync(),
             )
-            .await?;
+            .await
+            {
+                Ok(val) => val,
+                Err(error) => match error.downcast::<ZeroQuoteReceived>() {
+                    Ok(_) => {
+                        bail!("The Monero seller you contacted can't trade monero at the moment")
```
Comment (binarybaron) [inline swap/src/bin/swap.rs:115]: Small thing but maybe replace this with something like `Seller's XMR balance is currently too low to initiate a swap, please try again later`. We know what caused the error so we can be a little more specific.

File: swap/src/asb/event_loop.rs [hunk 26]
```diff
@@ -319,13 +319,46 @@ where
...
+                price: ask_price,
+                min_quantity: bitcoin::Amount::ZERO,
+                max_quantity: bitcoin::Amount::ZERO,
+            });
+        }
+
+        if max_buy > max_bitcoin_for_monero {
+            tracing::warn!(
+                    "Your Max bitcoin to sell config: {} is bigger than your current balance of Monero available to sell {}",
```
Comment (binarybaron) [inline swap/src/asb/event_loop.rs:350]: Small thing but maybe replace this with something like this so it's a little more consistent with the other log message

`Your Monero balance is too low to initiate a swap with the maximum swap amount {} that you have specified in your config. You can at most swap {}`

File: swap/src/bin/swap.rs (hunk 25 above)
Comment (lescuer97) [reply swap/src/bin/swap.rs:115]: Much better message hahaha

File: swap/src/asb/event_loop.rs (hunk 26 above)
Comment (lescuer97) [reply swap/src/asb/event_loop.rs:350]: Done!

**Comment by lescuer97** [comment]
Just added a small test to check for checking the error message

**Review by thomaseizinger — APPROVED** [review]
Nice! Thank you for your patience and incorporating all the feedback :)

To finalise this, can you please:

- Add a changelog entry
- Squash all commits into one

**Comment by lescuer97** [comment]
Done! @thomaseizinger. No patience needed hahaha I learned a lot in the process

**Comment by thomaseizinger** [comment]
> this also fixes #963

If you add this to thr PR description, it will close it automatically!

**Comment by thomaseizinger** [comment]
Looks good! Kicked off CI. If it passes, feel free to merge.

bors d=lescuer97

**Comment by thomaseizinger** [comment]
> > this also fixes #963
> 
> If you add this to thr PR description, it will close it automatically!

Note: Only if you give it its own line like this:

Fixes https://github.com/comit-network/xmr-btc-swap/issues/939.
Fixes https://github.com/comit-network/xmr-btc-swap/issues/963.
